README.md
# ConsulConfigProvider

[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)
[![Hex pm](http://img.shields.io/hexpm/v/consul_config_provider.svg?style=flat)](https://hex.pm/packages/consul_config_provider)
[![hexdocs.pm](https://img.shields.io/badge/docs-latest-green.svg?style=flat)](https://hexdocs.pm/consul_config_provider/)

## Installation

The package can be installed by adding `consul_config_provider` to your list of dependencies in `mix.exs`:

```elixir
def deps do
  [
    {:consul_config_provider, "~> 0.2"},
    {:finch, "~> 0.12"}, # default implementation for http client
  ]
end
```

- Only add in finch if you want to use the default http behaviour and not define your own client

The docs can be found at [https://hexdocs.pm/consul_config_provider](https://hexdocs.pm/consul_config_provider).

## Sample Usage

- In `mix.exs`

```elixir
releases: [
  release_name: [
    include_executables_for: [:unix],
    applications: [
      runtime_tools: :permanent,
      app_name_here: :permanent
    ],
    config_providers: [
      {ConsulConfigProvider,
       %{prefix: "services/app_namespace_in_consul/v1", app_name: :app_name_here}}
    ]
  ]
],
```

### Transformer

- You can also implement an optional transformer behaviour to change the form of your configs
- This is helpful for interop with erlang modules that might have different opinions about things

```elixir
defmodule Example.Config do
  @behaviour ConsulConfigProvider.Transformer

  @impl true
  def transform({:erlkaf, [clients: [producer: [client_options: client_options]]]}) do
    default_client_options =
      Application.get_env(:erlkaf, :clients, [])
      |> Keyword.get(:producer, [])
      |> Keyword.get(:client_options, [])

    {:erlkaf,
     [
       clients: [
         producer: [
           type: :producer,
           client_options: Keyword.merge(default_client_options, client_options)
         ]
       ]
     ]}
  end

  @impl true
  def transform(config), do: config
end
```

Then in your configs:

```elixir
config :consul_config_provider, transformer_module: Example.Config
```

## Information

- This provider assumes the config name has a file extension which is either `.json`, `.yml`, or `.yaml` no other extensions are supported although PRs would be welcomed. If you do not follow this naming convention the provider will not work and throw.
- In the above the `prefix` is used for the keys path and can also be set with `CONSUL_PREFIX` (provide an empty string to the prefix if you wish to just use the env var)
- The `CONSUL_HOST` env var is used for the host to talk to consul and defaults to localhost
- The `CONSUL_PORT` env var is used for the port to talk to consul and defaults to 8500
- The http_module is dynamic and you can specify your own if you choose to do so. Just implement the HTTP behaviour and make your implementation return an `{:ok, json_body_binary_string}` as per the finch example which will be the fallback. You might have to deal with coercing the input keyword list for the finch arguments to support the client you are using, as well. You also need to set `config :consul_config_provider, :http_module, Client.YourClient` pointing to your client in your configs.
- Dependency-related configs are namespaced with their own application name while other configs use the input app_name for their namespace

## Release

- Increment [mix.exs project version](https://github.com/blueshift-labs/consul_config_provider/blob/master/mix.exs#L7)
- Update CHANGELOG.md
- Create a release in the [github ui](https://github.com/blueshift-labs/consul_config_provider/releases/new)
- Github actions will create a hex release and upload it


CHANGELOG.md
# Changelog

## v0.2.4 (2022-05-23)

- Switch from mojito to finch since mojito is deprecated
- bump lib versions
  - jason
  - ex_doc
  - stream_data
  - yaml_elixir
  - telemetry
  - mox

## v0.2.2

- hex release

## v0.2.1

- allow dry run

## v0.2.0

- Support for CONSUL_PREFIX
- Breaking changes
  - Changes how tranforms are done  

## v0.1.4

- Namespace modules by `ConsulConfigProvider` [#1](https://github.com/blueshift-labs/consul_config_provider/pull/1)
- Do not force compiling the request function [#2](https://github.com/blueshift-labs/consul_config_provider/pull/2)
- README.md updates

## v0.1.3

- Adding in ConsulConfigProvider.Transformer behaviour

## v0.1.2

- Updating README.md and recursive string_atoms

## v0.1.1

- Upgrading mojito version to 0.6.0 and refactoring

## v0.1.0

- Initial implementation


config/test.exs
import Config

config :consul_config_provider, :http_module, ConsulConfigProvider.HttpMock


test/test_helper.exs
ExUnit.start()

Mox.defmock(ConsulConfigProvider.HttpMock, for: ConsulConfigProvider.Http)


test/consul_config_provider_test.exs
defmodule ConsulConfigProviderTest do
  use ExUnit.Case
  use ExUnitProperties

  import Mox

  setup :verify_on_exit!

  property "load/2 returns the updated config" do
    check all(
            prefix <- string(:alphanumeric, min_length: 1),
            config_key <- string(:alphanumeric, min_length: 1),
            config_value <- string(:alphanumeric, min_length: 1),
            config_name <- string(:alphanumeric, min_length: 1)
          ) do
      encoded_value =
        %{config_key => config_value}
        |> Jason.encode!()
        |> Base.encode64()

      encoded_config =
        [
          %{
            "CreateIndex" => 12,
            "Flags" => 0,
            "Key" => "#{prefix}/#{config_name}.json",
            "LockIndex" => 0,
            "ModifyIndex" => 1,
            "Value" => encoded_value
          }
        ]
        |> Jason.encode!()

      expect(ConsulConfigProvider.HttpMock, :request, 2, fn args ->
        url = Keyword.get(args, :url)

        case String.contains?(url, "?keys=true") do
          true ->
            # initial call to get keys
            {:ok, "[\"#{prefix}/#{config_name}.json\"]"}

          false ->
            # other call(s) to get key values
            {:ok, encoded_config}
        end
      end)

      new_config = ConsulConfigProvider.load(
        [consul_config_provider: Application.get_all_env(:consul_config_provider)],
        %{prefix: prefix}
      )

      assert get_in(new_config, [
               String.to_atom(config_name),
               String.to_atom(config_key)
             ]) == config_value
    end
  end
end


mix.exs
defmodule ConsulConfigProvider.MixProject do
  use Mix.Project

  def project do
    [
      app: :consul_config_provider,
      version: "0.2.4",
      elixir: "~> 1.9",
      start_permanent: Mix.env() == :prod,
      deps: deps(),
      description: description(),
      package: package(),

      # Docs
      name: "Consul Config Provider",
      source_url: "https://github.com/blueshift-labs/consul_config_provider",
      homepage_url: "https://github.com/blueshift-labs/consul_config_provider",
      docs: docs()
    ]
  end

  # Run "mix help compile.app" to learn about applications.
  def application do
    [
      extra_applications: [:logger]
    ]
  end

  # Run "mix help deps" to learn about dependencies.
  defp deps do
    [
      {:telemetry, "~> 1.1"},
      {:ex_doc, "~> 0.28", only: :dev, runtime: false},
      {:jason, "~> 1.3"},
      {:mox, "~> 1.0", only: :test},
      {:stream_data, "~> 0.5", only: :test},
      {:yaml_elixir, "~> 2.9"},
      {:finch, "~> 0.12", optional: true}
    ]
  end

  defp description() do
    "A simple somewhat opinionated consul config provider for elixir 1.9+ releases."
  end

  defp package() do
    [
      name: "consul_config_provider",
      files: ~w(lib doc .formatter.exs mix.exs README* LICENSE* CHANGELOG.md),
      licenses: ["MIT"],
      links: %{"GitHub" => "https://github.com/blueshift-labs/consul_config_provider"},
      source_url: "https://github.com/blueshift-labs/consul_config_provider"
    ]
  end

  defp docs() do
    [
      # The main page in the docs
      main: "readme",
      source_ref: "master",

      # logo: "path/to/logo.png",
      extras: ["README.md", "CHANGELOG.md"]
    ]
  end
end


lib/client/finch.ex
defmodule ConsulConfigProvider.Client.Finch do
  @moduledoc "Http client using finch"
  @behaviour ConsulConfigProvider.Http
  @compile {:inline, request: 1}

  if match?({:module, _}, Code.ensure_compiled(Finch)) do
    @impl true
    def request(method: method, url: url) do
      {:ok, _} = Application.ensure_all_started(:telemetry)
      pid = Process.whereis(ConsulConfigDefaultHTTPClient)

      if pid == nil do
        Finch.start_link(name: ConsulConfigDefaultHTTPClient)
      end

      headers = [{"Content-Type", "application/json"}]
      pool_opts = %{default: [max_idle_time: 60_000, size: 10, count: 1]}
      finch_req = Finch.build(method, url, headers, nil, [pools: pool_opts])
      {:ok, %Finch.Response{body: body}} = Finch.request(finch_req, ConsulConfigDefaultHTTPClient)
      {:ok, body}
    end
  else
    @impl true
    def request(_request_opts) do
      {:error, :finch_not_compiled}
    end
  end
end


lib/consul_config_provider.ex
defmodule ConsulConfigProvider do
  @moduledoc "A consul config provider for Elixir 1.9+ assuming the configs have an extension in the key name"
  @behaviour Config.Provider
  @dialyzer {:nowarn_function, load: 2}

  @impl true
  def init(%{prefix: prefix} = consul_config) when is_binary(prefix) do
    consul_config
  end

  @impl true
  def load(config, %{prefix: prefix}) do
    host = System.get_env("CONSUL_HOST", "localhost")
    port = System.get_env("CONSUL_PORT", "8500") |> String.to_integer()
    prefix = System.get_env("CONSUL_PREFIX", prefix)
    keys_url = "http://#{host}:#{port}/v1/kv/#{prefix}?keys=true"

    http_module =
      get_in(config, [:consul_config_provider, :http_module])
      |> case do
        nil -> ConsulConfigProvider.Client.Finch
        mod -> mod
      end

    transformer_module = get_in(config, [:consul_config_provider, :transformer_module])

    {:ok, body} = http_module.request(method: :get, url: keys_url)

    consul_configs =
      body
      |> Jason.decode!()
      |> Enum.reduce([], fn path, acc ->
        if String.ends_with?(path, "/") do
          acc
        else
          [String.replace_leading(path, prefix <> "/", "") | acc]
        end
      end)
      |> Enum.map(
        &Task.async(fn ->
          get_consul_key(http_module, host, port, prefix, &1)
        end)
      )
      |> Enum.map(&Task.await/1)

    Enum.reduce(consul_configs, config, fn cfg, acc ->
      {key, value} =
        case transformer_module do
          nil -> cfg
          _ -> transformer_module.transform(cfg)
        end

      Config.Reader.merge(
        acc,
        [{key, value}]
      )
    end)
  end

  defp get_consul_key(http_module, host, port, prefix, key_name) do
    url = "http://#{host}:#{port}/v1/kv/#{prefix}/#{key_name}"
    {:ok, body} = http_module.request(method: :get, url: url)

    key_val =
      body
      |> Jason.decode!()
      |> hd()
      |> Map.get("Value")
      |> Base.decode64!()

    new_config =
      key_name
      |> Path.extname()
      |> case do
        ".json" ->
          key_val |> Jason.decode!() |> Map.to_list()

        ".yml" ->
          key_val |> YamlElixir.read_from_string!(maps_as_keywords: true, atoms: true)

        ".yaml" ->
          key_val |> YamlElixir.read_from_string!(maps_as_keywords: true, atoms: true)

        _ ->
          raise "unsupported config format"
      end
      |> string_atoms()

    config_prefix =
      key_name
      |> Path.rootname()
      |> String.to_atom()

    {config_prefix, new_config}
  end

  defp string_atoms(list, acc \\ [])

  defp string_atoms([], acc) do
    Enum.reverse(acc)
  end

  defp string_atoms([{str, val} | rest], acc) when is_binary(str) and is_list(val) do
    string_atoms(rest, [{String.to_atom(str), string_atoms(val)} | acc])
  end

  defp string_atoms([{str, val} | rest], acc) when is_binary(str) do
    string_atoms(rest, [{String.to_atom(str), val} | acc])
  end

  defp string_atoms([item | rest], acc) do
    string_atoms(rest, [item | acc])
  end
end
